### PR TITLE
HAI-134 and HAI-133: Added error handling and more implementation to service

### PR DESCRIPTION
**Pull request by tintotti**
Tests changed accordingly. 
Changed implementation be more aligned with HankeGeometriaController and HankeGeometriaService

### CODE REVIEW

**Review by umeetiusbaar — CHANGES_REQUESTED** [review]
If you think my comments need some work then do it but if you see that it is okay as it is now then go ahead and merge. There were nothing really big issues in my opinion.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeController.kt [hunk 1]
```diff
@@ -1,81 +1,86 @@
...
-        } else {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(HankeError.HAI1017)
+        }
+        return try {
             val hanke = hankeService.loadHanke(hankeId)
-            if (hanke == null) {
-                return ResponseEntity.notFound().build()//TODO, error handling
-            } else {
-                return ResponseEntity.ok(hanke)
+            ResponseEntity.status(HttpStatus.OK).body(hanke)
```
[the remark this answers is not in the document]
Comment (tintotti) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeController.kt:30]: Yes. Fixed.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeController.kt [hunk 2]
```diff
@@ -1,81 +1,86 @@
...
-            if (createdHanke == null || id == null) {
-                //ResponseEntity.status(500) //TODO: Error handling for saving failed
-                ResponseEntity.unprocessableEntity().body("saving failed")  //just returning something
-            } else {
-                ResponseEntity.ok(createdHanke.hankeId as String)
+    /**
+     * Add one hanke.
+     *  TODO: user from front?
+     *  TODO: validation for input
+     * This method will be called when we do not have id for hanke yet
```
[the remark this answers is not in the document]
Comment (tintotti) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeController.kt:62]: Yes. Comment wrong.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeController.kt [hunk 3]
```diff
@@ -1,81 +1,86 @@
...
+        logger.info { "Entering createHanke ${hanke?.toJsonString()}" }
 
-        } else {
+        if (hanke == null) {
+            return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(HankeError.HAI1017)
+        }
 
-            var createdHanke = hankeService.save(hanke)
-            var id = createdHanke?.hankeId
+        return saveHanke(hanke)
```
[the remark this answers is not in the document]
Comment (tintotti) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeController.kt:55]: It is, but there was so much duplicated code that I created a method for it in the class and from there is the call to service.

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeService.kt [hunk 4]
```diff
@@ -1,9 +1,11 @@
 package fi.hel.haitaton.hanke
 
+import org.springframework.context.annotation.Bean
+import org.springframework.stereotype.Component
 import org.springframework.stereotype.Service
 
 
-@Service
+@Component
```
[the remark this answers is not in the document]
Comment (tintotti) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeService.kt:8]: Changed back

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeServiceImpl.kt [hunk 5]
```diff
@@ -1,16 +1,39 @@
 package fi.hel.haitaton.hanke
 
+import mu.KotlinLogging
+import org.springframework.stereotype.Component
 import java.time.ZonedDateTime
 
-class HankeServiceImpl : HankeService {
+private val logger = KotlinLogging.logger { }
+
+@Component
```
[the remark this answers is not in the document]
Comment (tintotti) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeServiceImpl.kt:9]: Changed back

File: services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeServiceImpl.kt [hunk 6]
```diff
@@ -1,16 +1,39 @@
...
     override fun loadHanke(hankeId: String): Hanke? {
         // TODO
         return Hanke("", true, "Hämeentien perusparannus ja katuvalot", ZonedDateTime.now(), ZonedDateTime.now(), "", 1)
     }
 
     override fun save(hanke: Hanke): Hanke? {
-        // TODO
-        return Hanke("", true, "Hämeentien perusparannus ja katuvalot", ZonedDateTime.now(), ZonedDateTime.now(), "", 1)
+
+        if (hanke.hankeId.isNullOrEmpty()) {
```
[the remark this answers is not in the document]
Comment (tintotti) [reply services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/HankeServiceImpl.kt:21]: So far they are identical. Let's change them when there are differences.

File: services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/HankeControllerTest.kt [hunk 7]
```diff
@@ -20,16 +20,15 @@ class HankeControllerTest {
...
     lateinit var hankeService: HankeService
 
+    lateinit var hankeController: HankeController
+
     @BeforeEach
     fun setUp() {
         MockitoAnnotations.initMocks(this)
-
+        hankeController = HankeController(hankeService)
         Mockito.`when`(hankeService.loadHanke(mockedHankeId))
```
[the remark this answers is not in the document]
Comment (tintotti) [reply services/hanke-service/src/test/kotlin/fi/hel/haitaton/hanke/HankeControllerTest.kt:32]: You are right. Might confuse at some point and give wrong test results.

**Review by umeetiusbaar — APPROVED** [review]

**Review by muhassin — APPROVED** [review]

**Pull request merged**